tracing-opentelemetry/src/main/java/io/micronaut/tracing/opentelemetry/interceptor/TraceInterceptor.java

```diff
@@ -0,0 +1,218 @@
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
+import io.opentelemetry.instrumentation.api.util.ClassAndMethod;
+import jakarta.inject.Singleton;
+import org.reactivestreams.Publisher;
+
+import java.util.concurrent.CompletionStage;
+
+/**
+ * Implements tracing logic for <code>ContinueSpan</code> and <code>NewSpan</code>
+ * using the Open Tracing API.
+ *
+ * @author Nemanja Mikic
+ */
+@Singleton
+@Requires(beans = Tracer.class)
+@InterceptorBean({ContinueSpan.class, NewSpan.class})
+public class TraceInterceptor implements MethodInterceptor<Object, Object> {
+
+    private final Instrumenter<ClassAndMethod, Object> instrumenter;
+
+    /**
+     * Initialize the interceptor with tracer and conversion service.
+     *
+     * @param openTelemetry the openTelemetry
+     */
+    public TraceInterceptor(OpenTelemetry openTelemetry) {
+        instrumenter = new MicronautCodeTelemetryBuilder(openTelemetry).build();
+    }
+
+    @Override
+    public int getOrder() {
+        return InterceptPhase.TRACE.getPosition();
+    }
+
+    /**
+     * Logs an error to the span.
+     *
+     * @param context the span
+     * @param e       the error
+     */
+    public static void logError(Context context, Throwable e) {
+        Span.fromContext(context).recordException(e);
+    }
+
+    @Nullable
+    @SuppressWarnings("unchecked")
+    @Override
+    public Object intercept(MethodInvocationContext<Object, Object> context) {
+        boolean isContinue = context.hasAnnotation(ContinueSpan.class);
+        AnnotationValue<NewSpan> newSpan = context.getAnnotation(NewSpan.class);
+        boolean isNew = newSpan != null;
+        if (!isContinue && !isNew) {
+            return context.proceed();
+        }
+        Context currentContext = Context.current();
+
+        ClassAndMethod classAndMethod = ClassAndMethod.create(context.getDeclaringType(), context.getMethodName());
+
+        if (isContinue) {
+            InterceptedMethod interceptedMethod = InterceptedMethod.of(context);
+            try {
+                switch (interceptedMethod.resultType()) {
+                    case PUBLISHER:
+                        Publisher<?> publisher = interceptedMethod.interceptResultAsPublisher();
+                        if (publisher instanceof TracingPublisher) {
+                            return publisher;
+                        }
+                        return interceptedMethod.handleResult(
+                            new TracingPublisher(publisher, null, classAndMethod, currentContext) {
+
+                                @Override
+                                public void doOnSubscribe(@NonNull Context openTelemetryContext) {
+                                    tagArguments(context, openTelemetryContext);
+                                }
+
+                            });
+                    case COMPLETION_STAGE:
```

Comment:
Technically it should be possible to trace a future.

Reply:
Right now `interceptedMethod.resultType()` has only this values:

```
    enum ResultType {
        COMPLETION_STAGE, PUBLISHER, SYNCHRONOUS
    }
```